community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java

```diff
@@ -44,129 +61,243 @@ public CommunityLockClient( LockManagerImpl manager )
+                    localLocks.put( resourceId, resource );
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
             }
         }
+        finally
+        {
+            stateHolder.decrementActiveClients();
+        }
     }
 
     @Override
     public void releaseShared( Locks.ResourceType resourceType, long resourceId )
     {
-        PrimitiveLongObjectMap<LockResource> localLocks = localShared( resourceType );
-        LockResource resource = localLocks.get( resourceId );
-        if( resource.releaseReference() != 0)
+        if ( !stateHolder.incrementActiveClients() )
         {
             return;
         }
+        try
+        {
+            PrimitiveLongObjectMap<LockResource> localLocks = localShared( resourceType );
+            LockResource resource = localLocks.get( resourceId );
+            if ( resource.releaseReference() != 0 )
+            {
+                return;
+            }
 
-        localLocks.remove( resourceId );
+            localLocks.remove( resourceId );
 
-        manager.releaseReadLock( new LockResource( resourceType, resourceId ), lockTransaction );
+            manager.releaseReadLock( new LockResource( resourceType, resourceId ), lockTransaction );
+        }
+        finally
+        {
+            stateHolder.decrementActiveClients();
+        }
     }
 
     @Override
     public void releaseExclusive( Locks.ResourceType resourceType, long resourceId )
     {
-        PrimitiveLongObjectMap<LockResource> localLocks = localExclusive( resourceType );
-        LockResource resource = localLocks.get( resourceId );
-        if( resource.releaseReference() != 0)
+        if ( !stateHolder.incrementActiveClients() )
         {
             return;
         }
-        localLocks.remove( resourceId );
+        try
+        {
+            PrimitiveLongObjectMap<LockResource> localLocks = localExclusive( resourceType );
+            LockResource resource = localLocks.get( resourceId );
+            if ( resource.releaseReference() != 0 )
+            {
+                return;
+            }
+            localLocks.remove( resourceId );
 
-        manager.releaseWriteLock( new LockResource( resourceType, resourceId ), lockTransaction );
+            manager.releaseWriteLock( new LockResource( resourceType, resourceId ), lockTransaction );
+        }
+        finally
+        {
+            stateHolder.decrementActiveClients();
+        }
     }
 
     @Override
     public void releaseAll()
     {
+        if ( !stateHolder.incrementActiveClients() )
+        {
+            return;
+        }
+        try
+        {
+            releaseLocks();
+        }
+        finally
+        {
+            stateHolder.decrementActiveClients();
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        // closing client to prevent any new client to come
+        stateHolder.closeClient();
```

Comment:
yes please welcome to separate forseti pr, but corresponding files here will be updated as well. Will rebase after first of two will be merged.